### PR TITLE
LibPDF: Implement object streams and fix some parsing and rendering bugs

### PR DESCRIPTION
**Pull request by metalvoidzz**
PDF version 1.5 introduced the so-called "Cross-Reference Streams", which are an alternative to the standard xref tables. These allow referencing objects that are part of an *object stream*, which makes it possible to compress multiple objects into a single stream object.

This PR introduces both parsing of this new kind of xref table, as well as object streams.

To make this possible, I decoupled the parsing of PDF objects and values from the main logic of document parsing. Parsing functions can now operate on buffers other than the loaded PDF, in this case decompressed streams.

Edit: My yakstack kept increasing with every document I tested, I've included some bug fixes to this PR.

### CODE REVIEW

File: Userland/Libraries/LibPDF/DocumentParser.cpp [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+
+    return table;
+}
+
+PDFErrorOr<NonnullRefPtr<XRefTable>> DocumentParser::parse_xref_table()
+{
+    if (!m_reader.matches("xref")) {
+        // Since version 1.5, there may be a cross-reference stream instead
+        return parse_xref_stream();
+    }
```
Comment (mattco98) [inline Userland/Libraries/LibPDF/DocumentParser.cpp:364]: I wonder if we should store the version of the document and error here if the version is not >=1.5. To my knowledge this is the first thing in LibPDF that should be explicitly version gated, though I'm sure there are other places where this should be done that I just wasn't aware of.

Comment (metalvoidzz) [reply Userland/Libraries/LibPDF/DocumentParser.cpp:364]: I guess it depends on how correct we want to be. For documents under 1.5 that are simply corrupted, ```parse_xref_stream``` will fail anyway. If it's just an incorrect version number, we could still read it perfectly fine, or at least try to :^)

Comment (mattco98) [reply Userland/Libraries/LibPDF/DocumentParser.cpp:364]: I wonder what the spec says in regards to that. It's possible they mandate a failure for not-yet-supported features like this, and in that case an explicit error would be better. But if it doesn't mention it then yeah, I think the implicit failure is fine.

Comment (metalvoidzz) [reply Userland/Libraries/LibPDF/DocumentParser.cpp:364]: I couldn't find something related in the spec, but I tried a 1.5 document with an incorrect version number with some viewers (MuPDF, Acrobat on Android) and they opened it just fine.

**Review by mattco98 — APPROVED** [review]

**Comment by metalvoidzz** [comment]
Converting this to draft, since I've got some more fixes coming :^)

File: Userland/Libraries/LibPDF/ObjectDerivatives.h [hunk 2]
```diff
@@ -113,6 +113,9 @@ class DictObject final : public Object {
     template<typename... Args>
     bool contains(Args&&... keys) const { return (m_map.contains(keys) && ...); }
 
+    template<typename... Args>
+    bool contains_any_of(Args&&... keys) const { return (m_map.contains(keys) || ...); }
```
Comment (mattco98) [inline Userland/Libraries/LibPDF/ObjectDerivatives.h:117]: Might be a good idea to rename `contains` to `contains_all_of`, to be more explicit

**Review by mattco98 — APPROVED** [review]
LGTM, as long as you tested these changes with the PDFs in the repo (and preferably others with the features you implemented, though from past experience I know how hard it can be to find/create a PDF with a particular format).

**Pull request merged**